crypto/src/swap.rs

```diff
@@ -109,14 +368,38 @@ impl TryFrom<&[u8]> for SwapCiphertext {
     }
 }
 
-// TODO: ideally this would live in `component/dex/` or maybe a new subcrate including only
-// dex-related logic, but the former causes cyclic import issues
 #[derive(Debug, Clone)]
 pub struct TradingPair {
     pub asset_1: AssetId,
     pub asset_2: AssetId,
```

Comment:
Maybe these should be private, so we can enforce the invariant that `asset_1 < asset_2` on decode / construction?

I think it would make sense to have
```rust
impl TradingPair {
    pub fn asset_1() -> asset::Id;
    pub fn asset_2() -> asset::Id;
    /// Checks that asset_1 < asset_2 or errors
    pub fn new(asset_1: asset::Id, asset_2: asset::Id) -> anyhow::Result<Self>;
    /// Constructs the canonical representation of the provided pair.
    pub fn canonical_order_for(pair: (asset::Id, asset::Id)) -> Self;
```

This way, we can enforce the ordering constraint, and provide people with a way to construct the canonical representative of the pair. By making the `canonical_order_for` constructor take a tuple, it's harder to accidentally mix it up.

Reply:
Is there any reason `new` shouldn't behave like `canonical_order_for`? Is there any reason we'd want to fail to construct a `TradingPair`?

Reply:
Well, one reason would be if we're decoding a protobuf — I think we want to enforce that the serializations are canonical, since we may want to use the TradingPair as a lookup key in the database.  In general I think it makes sense to require that data we get from others be canonically encoded, and only use the order-independent constructor in special cases (like in pcli when we are trying to construct a TradingPair for some requested trade, or something).